SEIR/NPI/Stacked.py

```diff
@@ -1,58 +1,67 @@
 import pandas as pd
 
 from .base import NPIBase
 
 REDUCE_PARAMS = ["alpha", "r0", "gamma", "sigma"]
 
+
 class Stacked(NPIBase):
-    def __init__(self, *, npi_config, global_config, geoids, loaded_df = None):
-        super().__init__(npi_config)
+    def __init__(self, *, npi_config, global_config, geoids, loaded_df=None):
+        super().__init__(name=npi_config.name)
 
         self.start_date = global_config["start_date"].as_date()
         self.end_date = global_config["end_date"].as_date()
 
         self.geoids = geoids
+        self.reductions = {param: None for param in REDUCE_PARAMS}  # lazily construct: None means no reduction
+        self.reduction_params = []
 
-        # Gather parameter reductions
-        reduction_lists = {}
-        for param in REDUCE_PARAMS:
-            reduction_lists[param] = []
-
-        self.sub_npis = []
-
-        self.intervention_name = npi_config.name
-
-        for scenario in npi_config["scenarios"]:
+        # the confuse library's config resolution mechanism makes slicing the configuration object expensive; instead,
+        # just preload all settings
+        settings_map = global_config["interventions"]["settings"].get()
+        for scenario in npi_config["scenarios"].get():
             # if it's a string, look up the scenario name's config
-            if isinstance(scenario.get(), str):
-                scenario_npi_config = global_config["interventions"]["settings"][scenario.get()]
-                if not scenario_npi_config.exists():
+            if isinstance(scenario, str):
+                settings = settings_map.get(scenario)
+                if settings is None:
                     raise RuntimeError(f"couldn't find scenario in config file [got: {scenario}]")
             else:
                 # otherwise use the specified map as the config
-                scenario_npi_config = scenario
-
-            sub_npi = NPIBase.execute(npi_config=scenario_npi_config, global_config=global_config, geoids=geoids, loaded_df = loaded_df)
-            self.sub_npis.append(sub_npi)
-
-
-        self.reductions = {}
-        for param in REDUCE_PARAMS:
-            self.reductions[param] = 1 - functools.reduce(lambda a,b : a * (1 - b.getReduction(param)), self.sub_npis, 1)
+                settings = scenario
+
+            # via profiling: faster to recreate the confuse view than to fetch+resolve due to confuse isinstance checks
+            scenario_npi_config = confuse.RootView([settings])
+            sub_npi = NPIBase.execute(npi_config=scenario_npi_config, global_config=global_config, geoids=geoids,
+                                      loaded_df=loaded_df)
+            for param in REDUCE_PARAMS:
+                reduction = sub_npi.getReduction(param)
+                if reduction is not None:
+                    self.reductions[param] = (self.reductions[param] * (1 - reduction)
+                                              if self.reductions[param] is not None else reduction)
```

Comment:
Should this be 1- reduction?

Reply:
i simplified this calculation so it's much easier to understand (reductions can be scalars or matrices.)